File: external/greenX/GX-TimeFrequency/test/test_gx_minimax_grid.f90
! **************************************************************************************************
!  Copyright (C) 2020-2023 GreenX library
!  This file is distributed under the terms of the APACHE2 License.
!
! **************************************************************************************************!> Note, this is essentially a unit test but there's no fortran unit test framework
!> in place. As such, we have to a) run by python, b) write to file c) python parse data and assert.
!>
!> Cannot write python bindings as all grids/weights are allocated in the library
!> - This should be changed
program test_gx_minimax_grid
    use kinds, only: dp
    use gx_minimax, only: gx_minimax_grid
    implicit none

    integer :: n_mesh_points
    real(dp) :: e_transition_min, e_transition_max
    real(dp), allocatable :: tau_mesh(:), tau_weights(:)
    real(dp), allocatable :: freq_mesh(:), freq_weights(:)
    real(dp), allocatable :: cos_tau_to_freq_weights(:, :), cos_freq_to_tau_weights(:, :)
    real(dp), allocatable :: sinft_tau_to_freq_weights(:, :)
    real(dp) :: max_errors(3)
    real(dp) :: cosft_duality_error
    integer :: ierr
    character(300) :: file_name, dummy_char, tmp_path

    integer :: i, unit_id, n

    ! Pass file name from command line
    call get_command_argument(1, file_name)

    ! Read settings from file
    open(newunit=unit_id, file=trim(adjustl(file_name)))
    ! Rather than parse through file_name to get split into tmp_path/file
    ! just put the path in the input file and extract
    read(unit_id, *) tmp_path
    read(unit_id, *) dummy_char, n_mesh_points
    read(unit_id, *) dummy_char, e_transition_min
    read(unit_id, *) dummy_char, e_transition_max
    close(unit_id)

    ! Call the library API
    call gx_minimax_grid(n_mesh_points, e_transition_min, e_transition_max, &
                         tau_mesh, tau_weights, &
                         freq_mesh, freq_weights, &
                         cos_tau_to_freq_weights, cos_freq_to_tau_weights, &
                         sinft_tau_to_freq_weights, &
                         max_errors, cosft_duality_error, ierr)

    ! Extract tmp_path from file_name, where file_name = tmp_path / 'inputs.dat'
    n = index(trim(file_name), 'inputs.dat')
    tmp_path = file_name(1:n-1)

    ! Tau grid and weights
    open(newunit=unit_id, file = trim(adjustl(tmp_path))//'tau.dat')
    do i = 1, size(tau_mesh)
        write(unit_id, *) tau_mesh(i), tau_weights(i)
    end do
    close(unit_id)

    ! Frequency grid and weights
    open(newunit=unit_id, file = trim(adjustl(tmp_path))//'freq.dat')
    do i = 1, size(freq_mesh)
        write(unit_id, *) freq_mesh(i), freq_weights(i)
    end do
    close(unit_id)

    ! TODO Finish
    ! cos and sin transform weights
!    open(newunit=unit_id, file = trim(adjustl(tmp_path))//'cos_sin_weights.dat')
!    do i = 1, size(cos_tau_to_freq_weights)
!        write(unit_id, *) cos_tau_to_freq_weights(i, j), cos_freq_to_tau_weights(i, j), &
!                    sinft_tau_to_freq_weights(i, j)
!    end do
!    close(unit_id)

    ! TODO Add max_errors and cosft_duality_error as outputs

    ! Note, arrays are allocated in the library, and so must be explicitly
    ! deallocated by the caller
    deallocate(tau_mesh, tau_weights, freq_mesh, freq_weights, &
               cos_tau_to_freq_weights, cos_freq_to_tau_weights, sinft_tau_to_freq_weights)

end program test_gx_minimax_grid

File: external/greenX/GX-TimeFrequency/test/test_gx_minimax_grid.py
"""
! **************************************************************************************************
!  Copyright (C) 2020-2023 GreenX library
!  This file is distributed under the terms of the APACHE2 License.
!
! **************************************************************************************************

Inputs and assertions for test_gx_minimax_grid.f90

Would be better if one had python API to GX, and avoid the need for this intermediate binary.
Indeed, MUCH of the complication regarding running tests comes from this.

Issues and solutions:
---------------------
* Where should the python test write the input data for the fortran binary?
  - This is solved using tmp_path, which gives an absolute path.
"""
import numpy as np
from pathlib import Path
import pytest
import os

from pygreenx.run import BinaryRunner, BuildType


@pytest.fixture()
@pytest.mark.usefixtures("get_binary", "greenx_build_root")
def fortran_binary(get_binary, greenx_build_root):
    name = 'test_gx_minimax_grid.exe'
    _binary = get_binary(name)
    assert _binary is not None, f'{name} cannot be found in {greenx_build_root}'
    print(f'Binary source: {_binary}')
    return _binary


def mock_file(path, inputs_str):
    """ Write an inputs string to file.
    :param path:
    :param inputs_str:
    :return: Path object
    """
    file: Path = Path(path) / Path("inputs.dat")
    file.write_text(inputs_str)
    assert file.exists(), f"{file.name} does not exist"
    return file


class ETrans:
    def __init__(self, min, max):
        self.min = min
        self.max = max


# Note. This could be a fixture
def get_grids_and_weights(fortran_binary, n_mesh_points, e_trans: ETrans):

    # Fortran input file
    inputs_template = """
{}
n_mesh_points     {}
e_transition_min  {}
e_transition_max  {}
"""

    working_dir = os.getcwd()

    inputs_str = inputs_template.format(working_dir, n_mesh_points, e_trans.min, e_trans.max)
    file = mock_file(working_dir, inputs_str)

    # Run test
    runner = BinaryRunner(fortran_binary, BuildType.serial, args=[file.as_posix()])
    results = runner.run()
    assert results.success, f"Execution of {fortran_binary} failed"

    # Parse results
    tau_file = "tau.dat"
    freq_file = "freq.dat"

    return np.genfromtxt(tau_file), np.genfromtxt(freq_file)


# @pytest.mark.usefixtures("testdir")
def test_get_grids_and_weights(fortran_binary):
    """
    TODO(Maryam) Issue 24. Extend the test inputs, and include fringe cases.
    """
    # Test inputs
    n_mesh_points = 10
    e_trans = ETrans(2., 30.)

    tau_grid_weights, freq_grid_weights = get_grids_and_weights(fortran_binary,
                                                                n_mesh_points,
                                                                e_trans)

    ref_tau_grid_weights = np.array([[0.005918, 0.01525399],
                                     [0.03177845, 0.03688566],
                                     [0.08092057, 0.0622535],
                                     [0.15862351, 0.09469842],
                                     [0.27439677, 0.13947465],
                                     [0.44433573, 0.20467764],
                                     [0.69464678, 0.30283334],
                                     [1.06797564, 0.45564411],
                                     [1.63956744, 0.7122558],
                                     [2.58159534, 1.25940329]])

    ref_freq_grid_weights = np.array([[0.37397426, 0.75729004],
                                      [1.17910655, 0.87356340],
                                      [2.16745299, 1.12983421],
                                      [3.50267933, 1.57927297],
                                      [5.42267468, 2.32153469],
                                      [8.30517692, 3.54877702],
                                      [12.81707194, 5.68714236],
                                      [20.35677335, 9.94234857],
                                      [34.86933737, 21.24030777],
                                      [75.79619716, 80.07961331]])

    assert np.allclose(tau_grid_weights, ref_tau_grid_weights)
    assert np.allclose(freq_grid_weights, ref_freq_grid_weights)
